hoodie-common/src/main/java/com/uber/hoodie/common/table/log/HoodieCompactedLogRecordScanner.java

```diff
@@ -63,69 +68,100 @@
+                    HoodieTimeline.LESSER_OR_EQUAL)) {
+                //hit a block with instant time greater than should be processed, stop processing further
               break;
+            }
+            switch (r.getBlockType()) {
+              case AVRO_DATA_BLOCK:
+                log.info("Reading a data block from file " + logFile.getPath());
+                // If this is a avro data block, then merge the last block records into the main result
+                merge(records, lastBlocks);
+                // store the last block
+                lastBlocks.push(r);
+                break;
+              case DELETE_BLOCK:
+                log.info("Reading a delete block from file " + logFile.getPath());
+                String lastBlockInstantTime = lastBlocks.peek().getMetadata().stream()
+                      .filter(HoodieLogBlockMetadata.I_TIME).findFirst().get().getValue();
+                if(lastBlockInstantTime.equals(blockInstantTime)) {
+                  // 2 blocks (data and delete) written in the same batch, don't merge yet
+                } else {
+                  // Block with the keys listed as to be deleted
+                  // This is a delete block, so lets merge any records from previous data block
+                  merge(records, lastBlocks);
+                }
+                // store deletes so can be rolled back
+                lastBlocks.push(r);
+                break;
+              case COMMAND_BLOCK:
+                log.info("Reading a command block from file " + logFile.getPath());
+                // This is a command block - take appropriate action based on the command
+                HoodieCommandBlock commandBlock = (HoodieCommandBlock) r;
+                String targetInstantForCommandBlock = r.getMetadata().stream()
+                        .filter(HoodieLogBlockMetadata.T_I_TIME).findFirst().get().getValue();
+                switch (commandBlock.getType()) { // there can be different types of command blocks
+                  case ROLLBACK_PREVIOUS_BLOCK:
+                    // Rollback the last read log block
+                    // Get commit time from last record block, compare with targetCommitTime, rollback only if equal,
+                    // this is required in scenarios of invalid/extra rollback blocks written due to failures during
+                    // the rollback operation itself
+                    HoodieLogBlock lastBlock = lastBlocks.peek();
+                    if (lastBlock != null &&
+                            targetInstantForCommandBlock.contentEquals(lastBlock.getMetadata()
+                                    .stream().filter(HoodieLogBlockMetadata.I_TIME).findFirst().get().getValue())) {
+                      log.info("Rolling back the last log block read in " + logFile.getPath());
+                      lastBlocks.pop();
+                    } else {
+                      log.info("Invalid or extra rollback command block in " + logFile.getPath());
```

Comment:
Please throw a loud exception indicating expected usage, if we can't rollback .. for e.g multiple rollbacks.

Reply:
So this code is to handle the following scenario : 

1. upsert(..) -> failed partially
2. rollback(inflight) -> failed partially -> wrote some rollback blocks
3. rollback(inflight) -> completed -> wrote rollback blocks in addition to the ones in the failed rollback.
Now we have extra rollback blocks due to partial failures. The `targetInstantTime` helps to safe guard rolling back correct data blocks for the extra rollback blocks that are written. So essentially, the intention is to just log rather than throw a loud exception.

Reply:
Then, may be have this in `WARN` level, as opposed to INFO? 

How do we do error handling for cases where user attempts to do a nested rollback? That's where I was getting at..

Reply:
Yes, `WARN` sounds correct. 
So, essentially the above implementation guards against nested rollbacks silently (since `targetInstantTine` will prevent them and a warn message will be logged) but there is a correctness issue from the perspective of the client/user since the client expected nested rollbacks.
For now, I was hoping to handle this in the documentation.
The problem is that `HoodieWriteClient` doesn't stop someone from actually doing nested rollbacks and so by the time the CompactedRecordScanner kicks in (basically the read path) there is no way to undo the nested rollbacks. At this point, if we find there are nested rollbacks and if we throw an error, the dataset is basically unusable (unless we have other mechanisms to fix this)..

Reply:
makes sense. doc is fine for now..